### ISSUE TITLE
Support ES7 async/await proposal

### ISSUE THREAD

**Issue by gaearon** [eslint/eslint#1657]
I'm talking about https://github.com/lukehoban/ecmascript-asyncawait. It's an accepted proposal for ES7.

Because https://github.com/6to5/6to5 can already compile these, it would be nice to have support for it behind a flag.

I wouldn't file issues for other ES7 features yet but this one is a game changer. It spoiled me as a C# developer a few years ago and I can't wait to start using it again.

(Yes I understand this can be done with `yield`. Bear with me!)

**Comment by lo1tuma** [comment]
Duplicate oft #1566

**Comment by gaearon** [comment]
Whoops, pardon. Anywhere I can read why ES7 is no go?

**Comment by nzakas** [comment]
This really comes down to what the parser supports. I do think it's a bit early to lock anything in since the spec hasn't yet been finalized (or started, really). But if Espree adds experimental support then we can also experiment with it.

I think this will be a common enough ask that we should at least investigate once all ES6 stuff is done.

**Comment by gaearon** [comment]
Thanks. Parser level support is all I'm asking for. I'm not advocating linting rules around it.

**Comment by dfreeman** [comment]
For what it's worth, we'd love to see this as well once the rest of the ES6 stuff lands.

We're currently using a utility wrapper with generators to accomplish the same thing, but most of the rest of our toolchain is picking up support for the first class `async`/`await` syntax, so it would be fantastic to see ESLint/Espree accept it too.

**Comment by nzakas** [comment]
If you want it, consider adding support to Espree: https://github.com/eslint/espree.

We will get to it, but it's low priority for the core team based on getting the rest of ES6 done. You can make it go faster by contributing. :)

**Comment by nzakas** [comment]
Given the recent developments in the world of parsers, I think that using babel-eslint is the best way to play with cutting edge features like this. https://github.com/babel/babel-eslint

That means you don't have to wait for us, and we don't need to worry about constantly updating experimental features.

**Comment by gaearon** [comment]
:+1:

**Comment by Turbo87** [comment]
@nzakas since async/await has progress to stage 4 would it make sense to reopen this issue?

**Comment by btmills** [comment]
@Turbo87 we're working on supporting async/await in the parser - you can follow eslint/espree#287 for updates!

**Comment by Turbo87** [comment]
Awesome thanks! I guess that will only be supported in babel 3 then, correct?

**Comment by btmills** [comment]
@Turbo87 I don't fully understand the question.... Once we've had time to implement support for the new features, you should be able to use ESLint and the default Espree parser by themselves (without needing the babel-eslint parser) for async/await and any of the [other stage 4 proposals](https://github.com/tc39/proposals/blob/master/finished-proposals.md). Whether you need the babel compiler will depend on the features supported by the JS engine(s) you're targeting.

**Comment by Turbo87** [comment]
Sorry, I meant eslint 3 😆

**Comment by btmills** [comment]
Ah! Yes, that should be coming to ESLint v3.something once it's ready.

**Comment by iroy2000** [comment]
I'm on "eslint": "^3.2.2"   

Is that async await rule support already in place ?

**Comment by platinumazure** [comment]
@iroy2000 You need to upgrade to eslint@3.6.0 or later to use async/await rule support.

**Comment by iroy2000** [comment]
👍   Thanks

**Comment by MarkHerhold** [comment]
What's the rule to enable async/await? I'm getting `unexpected token function` when doing:
```js
async function create() {
}
```

**Comment by ljharb** [comment]
@MarkHerhold set your `ecmaVersion` to `2017`.

**Comment by platinumazure** [comment]
@MarkHerhold You need to set ecmaVersion to 8 under parserOptions in configuration, or do the equivalent in CLI.

Configuration (JSON):

```json
{
    "parserOptions": {
        "ecmaVersion": 8 // or 2017
    },
    "rules": {}
    }
}
```

Via CLI:

```shell
eslint --parser-options="ecmaVersion:8"
```

**Comment by MarkHerhold** [comment]
@platinumazure That worked, thanks!

**Comment by oshalygin** [comment]
I get that ESLint supports ecmaVersion `8` but is that the right thing to do?  Should it just be `2017`?

Wildly confusing especially considering the convention is the year.

**Comment by not-an-aardvark** [comment]
@oshalygin It supports both.

**Comment by oshalygin** [comment]
@not-an-aardvark sorry I just edited, I get that it does, I'm just wondering if its proper for ESLint to support both versions, since the canonical version is `2017`.

I'm sure you guys discussed this in another thread, any chance you can reference it so I can read a little more about that decision?

**Comment by not-an-aardvark** [comment]
See https://github.com/eslint/eslint/issues/6750

**Comment by oshalygin** [comment]
Thanks!! @not-an-aardvark

**Comment by dcsan** [comment]
I still get an error with the above settings

  12:23  error  Parsing error: The keyword 'await' is reserved

.eslintrc.json  is as below. I also tried "ecmaVersion": 2017 | 7 | 8


```

  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
```

![image](https://cloud.githubusercontent.com/assets/514002/24435752/c4949afa-1469-11e7-9484-ef227001f9a3.png)

➜  stt git:(master) ✗ node node_modules/eslint/bin/eslint.js -v
v3.18.0

**Comment by not-an-aardvark** [comment]
Your syntax is invalid -- `await` can only be used in an async function.

You should replace your code with

```js
decode: async function(opts) {
  const mp3Stream = await (msg).readyStream()
  // ...
}
```

**Comment by dcsan** [comment]
thanks, however this also errors:

```
async function stt(opts) {
  debug("stt", opts)
  return("done")
}
```

  12:7  error  Parsing error: Unexpected token function

**Comment by not-an-aardvark** [comment]
That code parses fine for me. If you're still having trouble, please ask for help on our [gitter channel](https://gitter.im/eslint/eslint).

**Comment by jwalton** [comment]
This is definitely still failing for me:

```
$ mkdir t
$ cd t
$ npm install eslint
$ echo "async function foo() {}" > test.js
$ ./node_modules/.bin/eslint --init
? How would you like to configure ESLint? Answer questions about your style
? Are you using ECMAScript 6 features? Yes
? Are you using ES6 modules? Yes
? Where will your code run? Node
? Do you use JSX? No
? What style of indentation do you use? Spaces
? What quotes do you use for strings? Double
? What line endings do you use? Unix
? Do you require semicolons? Yes
? What format do you want your config file to be in? JavaScript

$ ./node_modules/.bin/eslint test.js

/Users/jwalton/t/test.js
  1:7  error  Parsing error: Unexpected token function

✖ 1 problem (1 error, 0 warnings)
```

**Comment by BlueRaja** [comment]
```
return async () => ...
return async function () { ... }
```
Both of these cause eslint errors in the latest version

**Comment by not-an-aardvark** [comment]
You need to specify `ecmaVersion: 8` in a config file to enable async function parsing. For more details, please read our [configuration docs](http://eslint.org/docs/user-guide/configuring#specifying-parser-options).

**Comment by platinumazure** [comment]
For what it's worth, I was able to reproduce some of the reported issues on
this thread using the espree demo *and* using the ES8 setting there. I
think this deserves a closer look.

**Comment by not-an-aardvark** [comment]
I was unable to reproduce the issues. All of the following parse without errors:

```js
require('espree').parse(`async function stt(opts) {
  debug("stt", opts)
  return("done")
}`, { ecmaVersion: 8 });

require('espree').parse('return async () => foo', { ecmaVersion: 8, ecmaFeatures: { globalReturn: true } });

require('espree').parse('return async function() { foo }', { ecmaVersion: 8, ecmaFeatures: { globalReturn: true } })
```

**Comment by BlueRaja** [comment]
@not-an-aardvark Yes I have `ecmaVersion: 2017` specified (which is the same as `8`), and code like this works fine:
```
async function blah() { 
  await whatever();
  ...
}
```
However the following fails with the error `error  Parsing error: Unexpected token function`:
```
describe('Some Mocha Test', () => {
  let target, context, exampleGame;
  ...
  const checkAttributes = function (callback) {
    return async function () { // Error here
      return callback(await target.getAttributesAsync(exampleGame, context));
    };
  };
```
and the following fails with the error `error  Parsing error: Unexpected token =>`
```
describe('Another Mocha Test', () => {
  let target, exampleSku;
  ...
  it('does something', async () => // Error here
    assert.equal(await target.getAttributesAsync(exampleSku), exampleSku));
});
```

**Comment by not-an-aardvark** [comment]
I can't reproduce this issue, it works fine for me.

If you're still having trouble, can you please fill out our [bug report template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md) and create a new issue, or ask on our [gitter channel](https://gitter.im/eslint/eslint)? Filling out the bug report template would make it much easier to help debug your problem.
